python/hail/methods/statgen.py

```diff
@@ -1721,25 +1723,110 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+        measure relatedness.
+    path : :obj:`str` or :obj:`None`
+        A temporary directory to store intermediate matrices. Storing the matrices
+        to a file system is necessary for reliable execution of this method.
+    block_size : :obj:`int`
+        the side length of the blocks of the block-distributed matrices; this
+        should be set such that at least three of these matrices fit in memory
+        (in addition to all other objects necessary for Spark and Hail).
+    min_kinship : :obj:`float`
+        Pairs of samples with kinship lower than ``min_kinship`` are excluded
+        from the results.
+    statistics : :obj:`str`
+        the set of statistics to compute, `phi` will only compute the
+        kinship statistic, `phik2` will compute the kinship and
+        identity-by-descent two statistics, `phik2k0` will compute the
+        kinship statistics and both identity-by-descent two and zero, `all`
+        computes the kinship statistic and all three identity-by-descent
+        statistics.
 
+    Returns
+    -------
+    :class:`.Table`
+        A :class:`.Table` mapping pairs of samples to estimations of their
+        kinship and identity-by-descent zero, one, and two.
+
+        The fields of the resulting :class:`.Table` entries are of types: `i`:
+        ``ds.colkey_schema``, `j`: ``ds.colkey_schema``, `kin`:
+        :py:data:`.tfloat64`, `k2`: :py:data:`tfloat64`, `k1`:
+        :py:data:`tfloat64`, `k0`: :py:data:`tfloat64`. The table is keyed by
+        `i` and `j`.
     """
-    require_col_key_str(dataset, 'pc_relate')
-    dataset = require_biallelic(dataset, 'pc_relate')
-    intstatistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
-    _, scores, _ = hwe_normalized_pca(dataset, k, False, True)
+    assert(scores._indices.source == ds)
+    # analyze('pc_relate_with_scores', scores, ds._col_indices)
+    ds = ds.annotate_cols(scores=scores)
+    ds = require_biallelic(ds, 'pc_relate_with_scores')
+    ds = require_unique_samples(ds, 'pc_relate_with_scores')
+    ds = ds.select_entries(GT=ds.GT,
+                           naa=ds.GT.num_alt_alleles().to_float64())
+    ds = ds.select_rows(
+        mean_gt=(agg.sum(ds.naa) /
+                 agg.count_where(functions.is_defined(ds.GT))))
+    mean_imputed_gt = functions.or_else(ds.naa, ds.mean_gt)
+    g = BlockMatrix.from_matrix_table(mean_imputed_gt,
+                                      path=path,
+                                      block_size=block_size)
+
+    pc_scores = (ds.index_cols('column_index').cols_table().collect())
+    pc_scores.sort(key=lambda x: x.column_index)
+    pc_scores = [x.scores for x in pc_scores]
+
+    int_statistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
+
+    col_keys_table = ds.col_keys()
+    col_keys_local = col_keys_table.collect()
+    col_keys_schema = col_keys_table.schema
+    col_keys_local_java = [col_keys_schema._convert_to_j(x) for x in col_keys_local]
```

Comment:
I don't know if this is still the right way to do this

Reply:
Can pcrelate in scala return a table with i,j ints which you then annotate to put in the keys?

Reply:
that seems like the right thing

Reply:
That seems reasonable, I'll let @jbloom22 decide what to do. There's a bit of a time crunch to get this out for Konrad.